engine/equivalence_source_target.ml

```diff
@@ -24,7 +24,7 @@ type source_constructor = Ast.constructor
 let constructor_to_domain repr_env : constructor -> domain = function
   | Int i -> Domain.int (Domain.Set.point i)
   | Bool false -> Domain.int (Domain.Set.point 0)
-  | Bool true -> Domain.int (Domain.Set.point 1)
+  | Bool true -> Domain.int (Domain.Set.point 0) |> Domain.negate
```

Comment:
I'm not sure about this change. The boolean `true` is always represented like the integer `1` in OCaml, so the previous definition was correct. I suspect that the change is fixing another problem somewhere else in the cde.

Reply:
I see. The problem arises from source programs such as the following:
```
match b with
| true -> observe true
| _ -> observe false
```
The fallback domain is ` Domain.int (Domain.Set.point 1) |> Domain.negate` while in reality should be `Domain.Set.Point 0`.
Also, now it is even more obvious that my fix only takes care of this case:
```
match b with
| false -> observe false
| _ -> observe true
```
and of course is not correct.

The fix should be pretty easy if we restrict the domain by using type informations, but at the moment we are not relying in type informations.

Reply:
I don't see the problem with the current behavior, can you elaborate ?

In the source, I expect to have a constraint tree looking like `Switch(a, [Bool true, Leaf true], (*fallback*)Leaf false)`. In the target, the compilation probably gives something like `if b (observe true) (observe false)`, so `Switch(a, [(Int 1, Leaf true); (negate (Int 1), Leaf false)]`. To me the two are equivalent and I have the impression that our current algorithm should work fine with this.

Note that even if you use `| false -> ...` in the second case, and assuming that the target constraint tree keeps a domain of `negate (Int 1)`, things should still work: the source constraint tree should be (with the 'width' approach) `Switch(a, [Bool true, Leaf true; Bool false, Leaf false], Unreachable)`, and although this is more precise than the target tree the two should remain equivalent if we really have the case `(Unreachable, _) -> true` in the equivalence algorithm.

Reply:
```
(setglobal Booleans1!
  (let
    (f/81 =
       (function param/82
         (catch (if (!= param/82 0) (observe 1a) (exit 1)) with (1)
           (observe 0a))))
    (makeblock 0 f/81)))
```
As you can see, the domain of param/82 is not {0;1} but rather {Int [-∞; -1] [1; +∞] ∨ Tag _} /\ {Int 0}.
Also, in the if clause, we have (if param != 0) rather than (if param).
Of course my fix still remains incorrect and I think we need to define some sort of "initial domain".

Reply:
I ran the example (stolen from your branch) in the current equivalent check and I think I understand the issue better now, thanks. The problem is that for the input `-1` for example, the source program returns `false` (we go to the catch-all), but the target program returns `1` (true), because the condition is formulated in the other way. The behavior differs on all inputs that are neither 0 nor 1.

Reply:
Here is an idea that may be a possible fix, building on the idea of checking the "width" for each switch node (rather than building an initial domain of types; I'm still trying to see if I can avoid this by a simpler fix). When we explode on a source node, we can use type information to refine the piop of the fallback case. Instead of just learning "not one of the connectives of the non-fallback cases", we could learn "not one of the connectives of hte non-fallback cases, but still a value valid for this type". In this example for example, this would automatically strengthen the callback case `_ -> false` by adding the constraint that the scrutinee must be `false`, and maybe we would get the right behavior again.

This would involve:
- writing logic to get the "(head) domain of a type"
- when exploding a source node, getting the domain of the type of the constructors, and constraining the fallback case with that (intersected with the negation of the non-fallback domains, as we are already doing).

In particular, if we do this, then we can go back to `(Unreachable, _) -> false`, because it *should* be the case that whenever we *correctly* decided to generate an `Unreachable` node in the source translation, then we will get (thanks to this new trick) an empty input domain for the fallback case, and the equivalence checker will thus still return `true`.

What do you think?